### PR TITLE
Config parameters

### PR DESCRIPTION
**Pull request by henryh2**
A pull request that adds a few parameters to the topograph config file and makes most payload arguments in the request JSON optional.

### CODE REVIEW

**Comment by henryh2** [comment]
I have identified that there is an issue caused by the merge of PR #12 that causes topograph to crash when requesting simulated block topology. Will also try to fix this issue in this PR.

File: README.md [hunk 1]
```diff
@@ -49,6 +49,22 @@ http:
...
   ssl: false
 
+# provider: the provider that topograph will use unless otherwise specified by the topology request.
+# Valid options include "aws", "oci", "gcp", "cw", "baremetal" or "test".
+# Default value if not specified with be "test"
+provider: "aws"
+
+# engine: the engine that topograph will use unless otherwise specified by the topology request.
+# Valid options include "slurm", "k8s", or "test".
+# Default value if not specified will be "test".
```
Comment (dmitsh) [inline README.md]: I think we don't have default value anymore. If the value is not provided here, and is missed in the payload, we are throwing an error, correct?

File: README.md [hunk 2]
```diff
@@ -49,6 +49,22 @@ http:
...
+engine: "slurm"
+
+# use_simulation: whether or not topograph should use topology simulation provided by toposim to fulfil topology requests.
+use_simulation: false
```
Comment (dmitsh) [inline README.md]: not in use anymore?

File: README.md [hunk 3]
```diff
@@ -49,6 +49,22 @@ http:
...
+
+# simulation_model_path: the file path to the model that, if simulation is enabled, is used to represent the topology of the simulated cluster
+simulation_model_path: ""
```
Comment (dmitsh) [inline README.md]: not in use anymore?

File: README.md [hunk 4]
```diff
@@ -133,6 +152,10 @@ Topograph offers three endpoints for interacting with the service. Below are the
     "creds": {
       "access_key_id": "id",
       "secret_access_key": "secret"
+    },
+    "params": {
+      "use_simulation": "false",
```
Comment (dmitsh) [inline README.md:150]: mocked_provider?

File: pkg/config/config.go [hunk 5]
```diff
@@ -31,6 +31,8 @@ import (
 type Config struct {
 	HTTP                    Endpoint          `yaml:"http"`
 	RequestAggregationDelay time.Duration     `yaml:"request_aggregation_delay"`
+	Provider                *string           `yaml:"provider"`
```
Comment (dmitsh) [inline pkg/config/config.go]: I would keep it "string", not "*string", as it is expected in most cases.

File: pkg/engines/k8s/labeler_test.go [hunk 6]
```diff
@@ -49,7 +50,7 @@ func TestApplyNodeLabels(t *testing.T) {
 		"Node306": {"topology.kubernetes.io/network-level-1": "xf946c4acef2d5939", "topology.kubernetes.io/network-level-2": "S1"},
 	}
 
-	err := NewTopologyLabeler().ApplyNodeLabels(context.TODO(), root, labeler)
+	err := NewTopologyLabeler().ApplyNodeLabels(context.TODO(), root.Vertices[common.ValTopologyTree], labeler)
```
Comment (dmitsh) [inline pkg/engines/k8s/labeler_test.go:53]: needs further discussion

File: README.md (hunk 2 above)
Comment (henryh2) [reply README.md]: Not used anymore.

File: README.md (hunk 3 above)
Comment (henryh2) [reply README.md]: Not used anymore.

File: README.md (hunk 1 above)
Comment (henryh2) [reply README.md]: Default values aren't necessary anymore.

File: README.md (hunk 4 above)
Comment (henryh2) [reply README.md:150]: Not used anymore, will be mocked_provider in a future PR.

File: pkg/config/config.go (hunk 5 above)
Comment (henryh2) [reply pkg/config/config.go]: Changed

File: pkg/engines/k8s/labeler_test.go (hunk 6 above)
Comment (henryh2) [reply pkg/engines/k8s/labeler_test.go:53]: Will resolve in future PR (very soon)

File: pkg/server/engine.go [hunk 7]
```diff
@@ -29,6 +30,9 @@ import (
 	"github.com/NVIDIA/topograph/pkg/utils"
 )
 
+var providerName string
```
Comment (dmitsh) [inline pkg/server/engine.go]: please avoid static variables. provider and engine might be passed in the payload.

File: pkg/server/grpc_client.go [hunk 8]
```diff
@@ -147,12 +149,22 @@ func toGraph(response *pb.TopologyResponse, cis []common.ComputeInstances) *comm
 		forest["extra"] = sw
 	}
 
-	root := &common.Vertex{
+	treeRoot := &common.Vertex{
```
Comment (dmitsh) [inline pkg/server/grpc_client.go:152]: Do we always return block format here?

File: pkg/server/http_server.go [hunk 9]
```diff
@@ -58,6 +58,10 @@ func InitHttpServer(ctx context.Context, cfg *config.Config) {
 			queue: utils.NewTrailingDelayQueue(processRequest, cfg.RequestAggregationDelay),
 		},
 	}
+
+	// Saves the provider and engine config parameters
+	providerName = cfg.Provider
```
Comment (dmitsh) [inline pkg/server/http_server.go]: we need to make validation here in set proper provider/engine in the topology request

File: pkg/server/http_server.go [hunk 10]
```diff
@@ -132,14 +136,14 @@ func validate(tr *common.TopologyRequest) error {
 	case common.ProviderAWS, common.ProviderOCI, common.ProviderGCP, common.ProviderCW, common.ProviderBM, common.ProviderTest:
 		//nop
 	case "":
-		return fmt.Errorf("missing provider name")
+		//nop, will default to provider set in config file
```
Comment (dmitsh) [inline pkg/server/http_server.go]: please revert

File: pkg/server/http_server.go [hunk 11]
```diff
@@ -132,14 +136,14 @@ func validate(tr *common.TopologyRequest) error {
...
 	default:
 		return fmt.Errorf("unsupported provider %s", tr.Provider.Name)
 	}
 
 	switch tr.Engine.Name {
 	case "":
-		return fmt.Errorf("missing engine name")
+		//nop, will default to engine set in config file
```
Comment (dmitsh) [inline pkg/server/http_server.go]: same here

**Review by dmitsh — CHANGES_REQUESTED** [review]

File: pkg/server/grpc_client.go (hunk 8 above)
Comment (henryh2) [reply pkg/server/grpc_client.go:152]: For the moment, will be fixed in future PR.

File: pkg/server/engine.go (hunk 7 above)
Comment (henryh2) [reply pkg/server/engine.go]: Fixed in last commit.

File: pkg/server/http_server.go (hunk 9 above)
Comment (henryh2) [reply pkg/server/http_server.go]: Fixed in last commit.

File: pkg/server/http_server.go (hunk 10 above)
Comment (henryh2) [reply pkg/server/http_server.go]: Fixed in last commit.

File: pkg/server/http_server.go (hunk 11 above)
Comment (henryh2) [reply pkg/server/http_server.go]: Fixed in last commit.

**Review by dmitsh — APPROVED** [review]

**Pull request merged**